### PR TITLE
IC09/IC10 Generate CE Estab Addressed PQs

### PR DESCRIPTION
**Pull request by LeoKHoward**
# Motivation and Context
Want to generate individual addressed initial contact questionnaires for CE Estabs in English and Welsh

# What has changed
New ActionTypes/PackCodes included
New unit/integration tests
New method for identifying Welsh QID type and generating a second UAC/QID pair

# How to test?
Build and run with other repos and then run ATs

# Links
https://trello.com/c/0JfdnR6q/795-ic09-ic10-generate-ce-ic-addressed-pqs-we-qids-required-13

### CODE REVIEW

**Comment by LeoKHoward** [comment]
[(IC09/IC10) Generate CE IC Addressed PQs: W&E QIDs required (13)](https://trello.com/c/0JfdnR6q/795-ic09-ic10-generate-ce-ic-addressed-pqs-we-qids-required-13)

File: src/main/java/uk/gov/ons/census/action/builders/QidUacBuilder.java [hunk 1]
```diff
@@ -45,9 +45,13 @@
   private static final int NUM_OF_UAC_QID_PAIRS_NEEDED_FOR_SINGLE_LANGUAGE = 1;
   private static final String WALES_IN_ENGLISH_QUESTIONNAIRE_TYPE = "02";
   private static final String WALES_IN_WELSH_QUESTIONNAIRE_TYPE = "03";
+  private static final String WALES_IN_ENGLISH_QUESTIONNAIRE_TYPE_CE_CASES = "22";
+  private static final String WALES_IN_WELSH_QUESTIONNAIRE_TYPE_CE_CASES = "23";
   private static final String UNKNOWN_COUNTRY_ERROR = "Unknown Country";
   private static final String UNEXPECTED_CASE_TYPE_ERROR = "Unexpected Case Type";
   public static final String HOUSEHOLD_INITIAL_CONTACT_QUESTIONNAIRE_TREATMENT_CODE_PREFIX = "HH_Q";
+  public static final String CE_INDIVIDUAL_INITIAL_CONTACT_QUESTIONNAIRE_TREATMENT_CODE_PREFIX =
```
Comment (NickGrantX) [inline src/main/java/uk/gov/ons/census/action/builders/QidUacBuilder.java]:
```suggestion
  public static final String CE_INITIAL_CONTACT_QUESTIONNAIRE_TREATMENT_CODE_PREFIX =
```

File: src/main/java/uk/gov/ons/census/action/builders/QidUacBuilder.java [hunk 2]
```diff
@@ -157,7 +171,9 @@ private UacQidLink createNewUacQidPair(Case linkedCase, String questionnaireType
   }
 
   private boolean isQuestionnaireWelsh(String treatmentCode) {
-    return (treatmentCode.startsWith(HOUSEHOLD_INITIAL_CONTACT_QUESTIONNAIRE_TREATMENT_CODE_PREFIX)
+    return ((treatmentCode.startsWith(HOUSEHOLD_INITIAL_CONTACT_QUESTIONNAIRE_TREATMENT_CODE_PREFIX)
```
Comment (NickGrantX) [inline src/main/java/uk/gov/ons/census/action/builders/QidUacBuilder.java:173]:
```suggestion
    return (treatmentCode.startsWith(HOUSEHOLD_INITIAL_CONTACT_QUESTIONNAIRE_TREATMENT_CODE_PREFIX
```

File: src/test/java/uk/gov/ons/census/action/builders/QidUacBuilderTest.java [hunk 3]
```diff
@@ -2,17 +2,14 @@
...
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.Assert.assertEquals;
-import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.ArgumentMatchers.*;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
-import static uk.gov.ons.census.action.builders.QidUacBuilder.HOUSEHOLD_INITIAL_CONTACT_QUESTIONNAIRE_TREATMENT_CODE_PREFIX;
-import static uk.gov.ons.census.action.builders.QidUacBuilder.WALES_TREATMENT_CODE_SUFFIX;
+import static org.postgresql.hostchooser.HostRequirement.any;
```
Comment (NickGrantX) [inline src/test/java/uk/gov/ons/census/action/builders/QidUacBuilderTest.java]: dodgy import. These are not the `any` droids you are looking for

Comment (AdamHawtin) [reply src/test/java/uk/gov/ons/census/action/builders/QidUacBuilderTest.java]: You need the mockito matchers `any` instead.

**Review by NickGrantX — CHANGES_REQUESTED** [review]
Few comments

File: src/test/java/uk/gov/ons/census/action/builders/QidUacBuilderTest.java (hunk 3 above)
Comment (AdamHawtin) [reply src/test/java/uk/gov/ons/census/action/builders/QidUacBuilderTest.java]: Ah, it's wildcard importing the argument matcher so this line can just be deleted

File: src/main/java/uk/gov/ons/census/action/builders/QidUacBuilder.java [hunk 4]
```diff
@@ -158,7 +171,8 @@ private UacQidLink createNewUacQidPair(Case linkedCase, String questionnaireType
 
   private boolean isQuestionnaireWelsh(String treatmentCode) {
     return (treatmentCode.startsWith(HOUSEHOLD_INITIAL_CONTACT_QUESTIONNAIRE_TREATMENT_CODE_PREFIX)
-        && treatmentCode.endsWith(WALES_TREATMENT_CODE_SUFFIX));
+        || treatmentCode.startsWith(CE_INITIAL_CONTACT_QUESTIONNAIRE_TREATMENT_CODE_PREFIX)
```
Comment (NickGrantX) [inline src/main/java/uk/gov/ons/census/action/builders/QidUacBuilder.java]: I have a concern that the boolean logic of this is slightly hard to work out:

`IF treatment code starts HH_Q OR treatment code starts CE_Q AND treatment code ends W THEN...`

It's not clear whether this always makes sure that the treatment code ends W. You should put the first part of the clause into its own clause.

Suggestion:

```
return ((treatmentCode.startsWith(HOUSEHOLD_INITIAL_CONTACT_QUESTIONNAIRE_TREATMENT_CODE_PREFIX)
        || treatmentCode.startsWith(CE_INITIAL_CONTACT_QUESTIONNAIRE_TREATMENT_CODE_PREFIX))
        && treatmentCode.endsWith(WALES_TREATMENT_CODE_SUFFIX));
```

**Review by NickGrantX — CHANGES_REQUESTED** [review]
Some boolean logic needs tightening up to make sure no bugs are introduced

File: src/main/java/uk/gov/ons/census/action/builders/QidUacBuilder.java (hunk 4 above)
Comment (LeoKHoward) [reply src/main/java/uk/gov/ons/census/action/builders/QidUacBuilder.java]: Done

**Review by NickGrantX — APPROVED** [review]
LGTM 👍

**Review by AdamHawtin — APPROVED** [review]

**Pull request merged**